samples/src/main/java/okio/samples/Throttler.java

```diff
@@ -0,0 +1,107 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okio.samples;
+
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import okio.Buffer;
+import okio.ForwardingSink;
+import okio.ForwardingSource;
+import okio.Sink;
+import okio.Source;
+
+/**
+ * Enables throttling of Sources and Sinks. Attach to this throttler via {@link #source(Source)} and
+ * {@link #sink(Sink)} and set the desired throughput via {@link #bytesPerTimePeriod(long, int,
+ * TimeUnit)}. Multiple Sources and Sinks can be attached to a single Throttler and they will be
+ * throttled as a group, where their combined throughput will not exceed the desired throughput. The
+ * same Source or Sink can be attached to multiple Throttlers and its throughput will not exceed the
+ * desired throughput of any of the Throttlers.
+ *
+ * <p>To clear any throttling, use {@link #clear()}.
+ */
+public final class Throttler {
+
+  private double limit = 0.0d;
+  private long last = System.currentTimeMillis();
+
+  /** Sets the desired max throughput. */
+  public void bytesPerTimePeriod(long byteCount, int duration, TimeUnit unit) {
+    Objects.requireNonNull(unit);
+    if (byteCount <= 0 || duration <= 0) {
+      throw new IllegalArgumentException("byteCount=" + byteCount + " duration=" + duration);
+    }
+
+    synchronized (this) {
+      this.limit = (double) byteCount / unit.toMillis(duration);
+    }
+  }
+
+  /** Clears the max throughput. */
+  public void clear() {
+    synchronized (this) {
+      this.limit = 0.0d;
+    }
+  }
+
+  /**
+   * Purposely delays until the caller of the method has taken a sufficient enough time so as not to
+   * exceed the desired throughput.
+   */
+  private void delay(long byteCount) throws InterruptedIOException {
+    synchronized (this) {
+      if (limit == 0.0d) return;
+
+      long required = (long) (byteCount / limit); // Total time required to take
+      long already = System.currentTimeMillis() - last; // Time since last request
+      long delay = required - already; // Time needs to be delayed
+
+      try {
+        if (delay > 0) Thread.sleep(delay);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new InterruptedIOException("interrupted");
+      }
+
+      last = System.currentTimeMillis();
+    }
+  }
+
+  /** Create a Source which honors this Throttler. */
+  public Source source(Source source) {
+    return new ForwardingSource(source) {
+      @Override
+      public long read(Buffer sink, long byteCount) throws IOException {
+        long read = super.read(sink, byteCount);
+        if (read > 0) delay(read);
```

Comment:
One thing I want to point out is that we are delaying after the read here because I thought it best to delay based on the amount of data that is actually read, not the requested byte count. For the Sink, we are delaying before the write so there is a little bit of an inconsistency. I can change either or neither, but I wanted to point it out specifically.

Reply:
Thanks for calling this out. I think it’s a nice design but I don’t think it’s what we want!

In particular we should assume somebody could be using the throttler to avoid consuming some expensive resource. Perhaps they’re sharing a narrow connection and they want to throttle one user to prioritize another.